File: src/video_io.py
import cv2

class VideoIO:
    def __init__(self, path):
        self.cap = cv2.VideoCapture(path)
        self.w = int(self.cap.get(3))
        self.h = int(self.cap.get(4))
        self.fps = int(self.cap.get(5))
        self.fc = int(self.cap.get(7))
        
    def read_all(self):
        frames = []
        while self.cap.isOpened():
            ret, frame = self.cap.read()
            if not ret: break
            frames.append(frame)
        return frames
    
    def read_range(self, start, end):
        self.cap.set(1, start)
        frames = []
        for _ in range(end - start):
            ret, frame = self.cap.read()
            if not ret: break
            frames.append(frame)
        return frames
    
    def write(self, path, frames):
        if not frames: return
        h, w = frames[0].shape[:2]
        writer = cv2.VideoWriter(path, cv2.VideoWriter_fourcc(*'mp4v'), self.fps, (w, h))
        for frame in frames:
            writer.write(frame)
        writer.release()
    
    def write_fps(self, path, frames, fps):
        if not frames: return
        h, w = frames[0].shape[:2]
        writer = cv2.VideoWriter(path, cv2.VideoWriter_fourcc(*'mp4v'), fps, (w, h))
        for frame in frames:
            writer.write(frame)
        writer.release()
    
    def extract_frame(self, frame_idx, output_path):
        self.cap.set(1, frame_idx)
        ret, frame = self.cap.read()
        if ret:
            cv2.imwrite(output_path, frame)
        return ret
    
    def get_frame_count(self):
        return self.fc
    
    def __del__(self):
        self.cap.release()

File: src/VideoIO.cpp
#include "../include/VideoIO.h"
#include <opencv2/imgcodecs.hpp>

using namespace cv;
using namespace std;

VideoIO::VideoIO(const std::string& path) {
    cap.open(path);
    if (!cap.isOpened()) {
        cerr << "Error: Cannot open video " << path << endl;
        exit(1);
    }
    w = (int)cap.get(CAP_PROP_FRAME_WIDTH);
    h = (int)cap.get(CAP_PROP_FRAME_HEIGHT);
    fps = (int)cap.get(CAP_PROP_FPS);
    fc = (int)cap.get(CAP_PROP_FRAME_COUNT);
}

VideoIO::~VideoIO() {
    cap.release();
}

vector<Mat> VideoIO::readAll() {
    vector<Mat> frames;
    Mat frame;
    while (cap.read(frame)) {
        frames.push_back(frame.clone());
    }
    return frames;
}

vector<Mat> VideoIO::readRange(int start, int end) {
    cap.set(CAP_PROP_POS_FRAMES, start);
    vector<Mat> frames;
    Mat frame;
    for (int i = 0; i < (end - start); i++) {
        if (!cap.read(frame)) break;
        frames.push_back(frame.clone());
    }
    return frames;
}

void VideoIO::write(const std::string& path, const vector<Mat>& frames) {
    if (frames.empty()) return;
    int h = frames[0].rows;
    int w = frames[0].cols;
    VideoWriter writer(path, VideoWriter::fourcc('m','p','4','v'), fps, Size(w, h));
    for (const auto& frame : frames) writer.write(frame);
    writer.release();
}

void VideoIO::writeFps(const std::string& path, const vector<Mat>& frames, int newFps) {
    if (frames.empty()) return;
    int h = frames[0].rows;
    int w = frames[0].cols;
    VideoWriter writer(path, VideoWriter::fourcc('m','p','4','v'), newFps, Size(w, h));
    for (const auto& frame : frames) writer.write(frame);
    writer.release();
}

bool VideoIO::extractFrame(int frameIdx, const std::string& outputPath) {
    cap.set(CAP_PROP_POS_FRAMES, frameIdx);
    Mat frame;
    if (cap.read(frame)) {
        imwrite(outputPath, frame);
        return true;
    }
    return false;
}
